torch/csrc/distributed/c10d/default_comm_hooks.cpp

```diff
@@ -23,7 +28,12 @@ c10::intrusive_ptr<c10::ivalue::Future> FP16CompressCommHook::runHook(
   compressed_tensor /= state_->getSize();
   std::vector<at::Tensor> tensors = {compressed_tensor};
 
-  auto allreduce_fut = state_->allreduce(tensors)->getFuture();
+  auto allreduce_fut =
+      ops::allreduce(
+          c10::intrusive_ptr<ProcessGroup>::unsafe_reclaim_from_nonowning(
+              state_),
```

Comment:
What's going on here?

Reply:
I'm trying to convert a raw pointer to an intrusive_ptr. Is this the way to do so?

Reply:
curious, who owns this PG instance? I assume it is owned by Python PG object? If that's the case, will this mess up the refcnt. What happens when this tmp intrusive ptr exits scope?

Reply:
I think PG are normally owned by a Python object. AllReduceCommHook somehow holds a ProcessGroup* instead of intrusive_ptr<ProcessGroup>. Therefore, I need to convert the raw pointer to a intrusive_ptr.

I don't believe this will mess up the refcnt. However, I actually think it's better to replace `class AllReduceCommHook : public CppCommHookInterface<ProcessGroup*>` with `class AllReduceCommHook : public CppCommHookInterface<intrusive_ptr<ProcessGroup>>`. What do you think?

Reply:
> I don't believe this will mess up the refcnt. 

If we create an intrusive ptr from the raw ptr, does this mean we have two separate entities tracking refcnt for the same raw ptr separately? One is the Python object, and another is this intrusive ptr?



> I actually think it's better to replace class AllReduceCommHook : public CppCommHookInterface<ProcessGroup*> with class AllReduceCommHook : public CppCommHookInterface<intrusive_ptr<ProcessGroup>>. What do you think?

Yep, this does sounds better to me.

Reply:
I think for intrusive_ptr the refcnt is stored in the object (ProcessGroup) itself. Intrusive_ptr is just a way to increment/decrement the refcnt. So it shouldn't matter.

Let me make a follow up patch on changing `class AllReduceCommHook : public CppCommHookInterface<ProcessGroup*>` to `class AllReduceCommHook : public CppCommHookInterface<intrusive_ptr<ProcessGroup>>`.

Reply:
Got it. Can we also add a comment for this in the code? Thank you!